src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs

```diff
@@ -307,34 +307,35 @@ public override object Clone()
                 this.logger.LogTrace("(-)[IS_IBD]");
                 return;
             }
-            
+
             //uint32_t nFetchFlags = GetFetchFlags(pfrom, chainActive.Tip(), chainparams.GetConsensus());
 
-            var send = new GetDataPayload();
-            foreach (InventoryVector inv in invPayload.Inventory.Where(inv => inv.Type.HasFlag(InventoryType.MSG_TX)))
-            {
-                //inv.type |= nFetchFlags;
+            var inventoryTxs = invPayload.Inventory.Where(inv => inv.Type.HasFlag(InventoryType.MSG_TX));
 
-                // TODO: This is incorrect, in blocks only mode we should just add to known inventory but not relay
-                if (this.isBlocksOnlyMode)                 
-                    this.logger.LogInformation("Transaction ID '{0}' inventory sent in violation of protocol peer '{1}'.", inv.Hash, peer.RemoteSocketEndpoint);
+            lock (this.lockObject)
+            {
+                foreach (var inv in inventoryTxs)
+                {
+                    this.filterInventoryKnown.Add(inv.Hash);
+                }
+            }
 
-                if (await this.orphans.AlreadyHaveAsync(inv.Hash))
+            var send = new GetDataPayload();
+            foreach (var inv in inventoryTxs)
+            {
+                if (await this.orphans.AlreadyHaveAsync(inv.Hash).ConfigureAwait(false))
                 {
                     this.logger.LogDebug("Transaction ID '{0}' already in orphans, skipped.", inv.Hash);
                     continue;
                 }
 
-                send.Inventory.Add(inv);
-            }
-
-            // add to known inventory
-            lock (this.lockObject)
-            {
-                foreach (InventoryVector inventoryVector in send.Inventory)
+                if (this.isBlocksOnlyMode)
```

Comment:
Previously this was the first condition in the for loop and it probably makes sense because you just want to log all transactions and do nothing otherwise. You will always execute the orphans and at the end you will just log it anyway. Is there any side effect of the orphan call?

Reply:
There is some rejects logic inside that AlreadyHaveAsync() which I'm not sure we want to skip.

Reply:
The argument here is that unless the orphan function has any side effects, which means it can only decide whether to continue not, then the order of these conditions here should be reversed

Reply:
I do agree with that in theory the inv is being sent out of protocol so should be ignored but this is the part of the orphan function I'm concerned about not executing:
```
 if (this.chain.Tip.HashBlock != this.hashRecentRejectsChainTip)
 {
      // If the chain tip has changed previously rejected transactions
      // might be now valid, e.g. due to a nLockTime'd tx becoming valid,
      // or a double-spend. Reset the rejects filter and give those
      // txs a second chance.
      this.logger.LogTrace("Executing task to clear rejected transactions.");
      this.hashRecentRejectsChainTip = this.chain.Tip.HashBlock;
      this.recentRejects.Clear();
}
```
In BTC core they also always execute this orphan functionality. Not sure if this is intentional or just a side effect of the way they wrote it.